tests/pytorch/test_numerics.py

```diff
@@ -675,3 +675,101 @@ def test_layernorm_linear_accuracy(dtype, bs, model):
+    static_target = torch.randn(config.seq_len, bs, config.hidden_size, device='cuda', dtype=dtype)
+
+    real_input = torch.rand_like(static_input)
+    real_target = torch.rand_like(static_target)
+
+    if graph:
+        # Pre graph capture warmup in a separate stream.
+        s = torch.cuda.Stream()
+        s.wait_stream(torch.cuda.current_stream())
+        with torch.cuda.stream(s):
+            for _ in range(3):
+                optimizer.zero_grad(set_to_none=True)
+                out = block(static_input)
+                loss = loss_fn(out, static_target)
+                loss.backward()
+                optimizer.step()
+        torch.cuda.current_stream().wait_stream(s)
+
+        # Capture.
+        g = torch.cuda.CUDAGraph()
+        optimizer.zero_grad(set_to_none=True)
+        with torch.cuda.graph(g):
+            static_output = block(static_input)
+            static_loss = loss_fn(static_output, static_target)
+            static_loss.backward()
+            optimizer.step()
+
+        # Fills the graph's input memory with new data to compute on
+        with torch.no_grad():
+            static_input.copy_(real_input)
+            static_target.copy_(real_target)
+        g.replay()
+    else:
+        with torch.no_grad():
+            static_input.copy_(real_input)
+            static_target.copy_(real_target)
+        optimizer.zero_grad(set_to_none=True)
+        static_output = block(static_input)
+        loss = loss_fn(static_output, static_target)
+        loss.backward()
+        optimizer.step()
+
+    torch.cuda.synchronize()
+
+    grads = [static_input.grad]
+    for p in block.parameters():
+        if p.requires_grad:
+            grads.append(p.grad)
+
+    with torch.no_grad():
+        output = static_output.clone()
+    return output, grads
+
+
+@pytest.mark.parametrize("dtype", param_types)
+@pytest.mark.parametrize("bs", batch_sizes)
+@pytest.mark.parametrize("model", model_configs.keys())
+def test_gpt_cuda_graph(dtype, bs, model):
+    config = model_configs[model]
+
+    sigma = 0.023
+    init_method = init_method_normal(sigma)
+    output_layer_init_method = scaled_init_method_normal(sigma, config.num_layers)
+
+    block = (
+        TransformerLayer(
+            config.hidden_size,
+            4 * config.hidden_size,
+            config.num_attention_heads,
+            layernorm_epsilon=config.eps,
+            init_method=init_method,
+            output_layer_init_method=output_layer_init_method,
+            hidden_dropout=0.1,
+            attention_dropout=0.1,
+            kv_channels=config.embed,
+            apply_residual_connection_post_layernorm=False,
+            output_layernorm=False,
+        )
+        .to(dtype=dtype)
+        .cuda()
+    )
+
+    out, _ = _test_sanity_e2e_cuda_graph(block, bs, dtype, config, False)
+    graph_out, _ = _test_sanity_e2e_cuda_graph(block, bs, dtype, config, True)
```

Comment:
I think we should run with two separate copies of `block`. We're applying SGD steps, so I'd expect each run to perturb the params.